toucan_connectors/snowflake_common.py

```diff
@@ -70,58 +95,168 @@ def _execute_query(
+            return DataSlice(self.data, self.count)
 
     def _fetch_data(
         self,
         c,
         data_source: SfDataSource,
         offset: Optional[int] = None,
         limit: Optional[int] = None,
+        get_row_count: bool = False,
     ) -> pd.DataFrame:
-        df = self._execute_query(c, data_source.query, data_source.parameters, offset, limit)
-        return df
+        toto = self._execute_parallelized_queries(
+            c, data_source.query, data_source.parameters, offset, limit, get_row_count
+        )
+        return toto.df
 
-    def retrieve_data(self, c, data_source: SfDataSource) -> pd.DataFrame:
-        return self._fetch_data(c, data_source)
+    def retrieve_data(
+        self, c, data_source: SfDataSource, get_row_count: bool = None
+    ) -> pd.DataFrame:
+        return self._fetch_data(c, data_source, get_row_count=get_row_count)
 
     def get_slice(
         self,
         c,
         data_source: SfDataSource,
         offset: Optional[int] = None,
         limit: Optional[int] = None,
+        get_row_count: bool = False,
     ) -> DataSlice:
-        df: pd.DataFrame = self._fetch_data(c, data_source, offset, limit)
+        result = self._fetch_data(c, data_source, offset, limit, get_row_count)
         if offset and limit:
-            result = df[offset : limit + offset]
-        else:
-            result = df
-        return DataSlice(result, len(result))
+            result = result[offset : limit + offset]
+        stats = DataStats(
+            execution_time=self.execution_time,
+            conversion_time=self.conversion_time,
+            total_returned_rows=len(result),
+            df_memory_size=result.memory_usage().sum(),
+        )
+
+        return DataSlice(
+            df=result,
+            input_parameters={
+                'limit': extract_limit(data_source.query),
+                'offset': extract_offset(data_source.query),
+            },
+            stats=stats,
+        )
 
     def get_warehouses(self, c, warehouse_name: Optional[str] = None) -> List[str]:
         query = 'SHOW WAREHOUSES'
         if warehouse_name:
             query = query + ' LIKE ' + warehouse_name
-        return [
-            warehouse['name'] for warehouse in self._execute_query(c, query) if 'name' in warehouse
-        ]
+        res = self._execute_query(c, query).to_dict().get('name')
+        return [warehouse for warehouse in res.values()] if res else []
 
     def get_databases(self, c, database_name: Optional[str] = None) -> List[str]:
         query = 'SHOW DATABASES'
         if database_name:
             query = query + ' LIKE ' + database_name
-        return [
-            database['name'] for database in self._execute_query(c, query) if 'name' in database
-        ]
+        res = self._execute_query(c, query).to_dict().get('name')
+        return [database for database in res.values()] if res else []
+
+    def __prepare_query(

```

Comment:
```suggestion
    def _prepare_query(
```

la convention en python c'est : 

- `_private_method` (i.e. je suis pas sensé être called hors du code de ma classe
- `__magic_method__`

Reply:
:+1: